### ISSUE TITLE
excalidraw CLI

### ISSUE THREAD

**Issue by eunjae-lee** [excalidraw/excalidraw#1261]
Hi there, 
I was wondering, what if we have a CLI for excalidraw?

```bash
excalidraw my-drawing.excalidraw
```

This will open a browser. If `my-drawing.excalidraw` already exists, it will load it.

When user clicks "save" button, it won't show the dialog but quietly save to `my-drawing.excalidraw`.

```bash
excalidraw my-drawing.excalidraw --export
```

with `--export` given, whenever user saves, the PNG version is exported as well.

```bash
excalidraw my-drawing.excalidraw --export images/my-drawing.png
```

---

I guess this can unblock new workflows, for example, for developers who want to have drawings in their git repository.

What do you think?

**Comment by tommywalkie** [comment]
Hello there, I'm currently trying to implement it in the following repo : [`tommywalkie/excalidraw-cli`](https://github.com/tommywalkie/excalidraw-cli).
A lot of CLI key features / refactoring / testing are still _work-in-progress_ but it already works for directories with `*.excalidraw` files.
#### Early concept demo
![demo](https://raw.githubusercontent.com/tommywalkie/excalidraw-cli/master/.github/assets/demo.gif)

I may consider opening another issue to export some useful Excalidraw renderer methods and constants (especially [`drawElementOnCanvas()`](https://github.com/excalidraw/excalidraw/blob/046c0818c5b39b78c70646b5f9a1c28f31787694/src/renderer/renderElement.ts#L86-L153)) so it can be easily integrated with this tool.

**Comment by j-f1** [comment]
Awesome @tommywalkie! When you feel like it’s more polished, would you be willing to perhaps collab with @MichaelDeBoey to use it for `gatsby-remark-embedder`? PR here: https://github.com/MichaelDeBoey/gatsby-remark-embedder/pull/122

**Comment by tommywalkie** [comment]
Hi @MichaelDeBoey @j-f1, thanks for your response, I would be glad to collab once I polished the CLI. Initially, I was planning to make it work for Rollup (for a Sapper/Svelte blog) but implementing it as a Gatsby plugin would be awesome!

**Comment by tommywalkie** [comment]
For reference : the experimental `excalidraw-cli` is now available for testing on npm as [`@tommywalkie/excalidraw-cli`](https://www.npmjs.com/package/@tommywalkie/excalidraw-cli) (this is scoped with my username, so you can still opt for your own implementation, if needed)
```bash
# Install
npm install -g @tommywalkie/excalidraw-cli

# Start testing the CLI
excalidraw-cli --help
excalidraw-cli <INPUT> <OUTPUT>
```
#### Features :
- Process **file** as entry, export as PNG (defaults to `${cwd}/${inputFileName}.png` if `<OUTPUT>` is undefined)
- Process **directory** as entry, export as PNG files in directory path (defaults to `${cwd}/` if `<OUTPUT>` is undefined)

Still a lot of things to do, but works fine so far.

**Comment by tommywalkie** [comment]
The experimental CLI app has been polished and is now close to a stable release.
#### Demo (`0.3.2`)
![demo](https://raw.githubusercontent.com/tommywalkie/excalidraw-cli/master/.github/assets/demo.gif)

```bash
$ excalidraw-cli --help
Parses Excalidraw JSON schemas into PNGs

USAGE
  $ excalidraw-cli [INPUT] [OUTPUT]

ARGUMENTS
  INPUT   [default: {cwd}] Excalidraw file path / directory path
  OUTPUT  [default: {cwd}] Output PNG file path / directory path

OPTIONS
  -h, --help     show CLI help
  -q, --quiet    disable console outputs
  -v, --version  show CLI version
```

@MichaelDeBoey @j-f1 I can start working on integrating it with Gatsby, I may have some questions, I may be wrong but considering this CLI tool can handle JSON data / Excalidraw files as input, wouldn't be better to create a Remark transformer instead ? 

Or maybe we can add some JSON API endpoint to the Excalidraw app itself so the cli tool can handle it and use it in `gatsby-remark-embedder` ?

**Comment by MichaelDeBoey** [comment]
@tommywalkie What `gatsby-remark-embedder` is doing is given the URL for a certain thing (tweet/video/sandbox/...), transforming that link to the proper embed code.

So what I want to do is provide you with a link and get the correct embed code or `svg` back

**Comment by tommywalkie** [comment]
@MichaelDeBoey I think my question was poorly worded, I was wondering _how to actually get the JSON_ from Excalidraw, using some URL link. 
But now, after some researches, I became aware of `https://json.excalidraw.com/api/v2/` and the exported method `importFromBackend` which may help with my issue.
https://github.com/excalidraw/excalidraw/blob/482fa2d90f9f56ba9fa0dd2cc458f3af232c5598/src/data/index.ts#L235-L238
If this is what I think, I can add "_Handle Excalidraw URLs as input_" as a new CLI feature which then would provide a lightweight easy-to-use alternative to `puppeteer` in https://github.com/MichaelDeBoey/gatsby-remark-embedder/pull/122

![Untitled-2020-06-22-1104(1)](https://user-images.githubusercontent.com/16148332/85275842-361ef200-b481-11ea-9388-9eaa9e1d7adf.png)

**Comment by MichaelDeBoey** [comment]
@tommywalkie After talking to @j-f1 this is my understanding of how @excalidraw is working:
- I go to http://excalidraw.com/#json=id,key
- The app sends a request to http://json.excalidraw.com/api/v2/id
- The app gets back an encrypted blob
- The app decodes the encrypted blob with the `key`
So that's a problem for having some sort of [oembed](https://oembed.com) possibility.

The thing I'm doing in the `gatsby-remark-embedder` plugin is either resolve the url myself or doing an API call
Now that's always to the oembed endpoint, but I can use another endpoint too of course

So I could use http://json.excalidraw.com, but then I have to do the decode of the blob myself

Another option would be to copy-paste the @excalidraw code (which is open source, right?), but that would cause problems each time you guys change something on your end so that's not the way we want to go.

So what I would like to have is having a lightweight way (this could be an `npm` package like `excalidraw-cli`) to pass the URL and get the decoded svg back.

**Comment by tommywalkie** [comment]
> Another option would be to copy-paste the @excalidraw code, but that would cause problems each time you guys change something on your end so that's not the way we want to go.

@MichaelDeBoey I understand the issue very well, I already opened https://github.com/excalidraw/excalidraw/issues/1780 because I have this exact problem on the rendering process because I don't have access to [`drawElementOnCanvas()`](https://github.com/excalidraw/excalidraw/blob/046c0818c5b39b78c70646b5f9a1c28f31787694/src/renderer/renderElement.ts#L86-L153) method, so I had to re-implement the renderer myself. 😅 It's doing fine currently but I wish I could rely on Excalidraw renderer directly for consistency, maybe I should open a PR about this.

**Comment by pomdtr** [comment]
Kroki-cli now support svg generation using excalidraw -> https://github.com/yuzutech/kroki-cli

**Comment by lipis** [comment]
I didn't follow the whole conversation.. but should we maybe create an official `excalidraw/excalidraw-cli` repo and take it from there? It seems that we need this tool.

**Comment by Timmmm** [comment]
Here's another one I just found: https://github.com/JRJurman/excalidraw-to-svg

Seems to do things using JSDOM to create a fake page with an export script. Probably the best approach.

There are a couple of bugs in it but it did work after I fixed them.

**Comment by dwelle** [comment]
Cool. Note that we will at some point be decoupling the renderer, UI, and the editor, so that you won't have to fake the DOM. But it's unclear when we'll be able to get to it.

**Comment by Timmmm** [comment]
That would be great. Looking forward to a WASM based implementation with a pluggable rendering API! :-P

In the mean time I finally wrote a solid tool to do it. Thanks to @JRJurman - probably would have given up trying to figure JSDom out with his code. I manage to simplify it quite a lot, fix various issues, and also add PDF export and some workarounds for font issues (e.g. embedding them).

Here it is:  👉  https://www.npmjs.com/package/excalidraw_export 👈 

    npm install --global excalidraw_export
    excalidraw_export foo.excalidraw

Please let me know if the `npm install --global excalidraw_export` command doesn't work! Annoying to test.

**Comment by godalming123** [comment]
I feel like this could be achieved using the web's [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API), rathor then forcing the user to download a package. It may also be easier to implement using this API.

**Comment by realazthat** [comment]
> I feel like this could be achieved using the web's [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API), rathor then forcing the user to download a package. It may also be easier to implement using this API.
> 
> > Offline functionality could then also be achieved using [Progressive Web Apps](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_appsl).
> 
> **Update:** I see that excalidraw already works offline using PWAs.
> 
> I don't see any reason to implement a CLI instead of using web APIs, using web APIs also allows for functionality on every OS.

Forgive my ignorance, how does one use a File System API or web API from another program or the command line?

**Comment by godalming123** [comment]
> How does one use a File System API or web API from another program or the command line?

I'm a bit confused about what you mean exactly.
- Firstly, the web file system API cannot be used from the command line, and is only designed for use within websites.
- Because the web's file system API can only be used within websites, this feature would have to be programmed into excalidraw itself. This would happen by editing the javascript (and possibly typescript) files within excalidraw to use the web's file system API to save the file without querying the user (MDN docs to the web's file system API are linked in my previous comment)

**Comment by realazthat** [comment]
> > How does one use a File System API or web API from another program or the command line?
> 
> I'm a bit confused about what you mean exactly

Oh hmm I thought you were responding to the original post, and suggesting that the Filesystem APIs and Web APIs can solve the issue, and I was confused.

> to save the file without querying the user.

To clarify, this is not the question of the OP. The OP is about being able to run the the entire `excalidraw => image` workflow from the command line, which would allow it to be used as a tool, like most such tools (e.g graphviz). Maybe I am misunderstanding your suggestion, my response is with the understanding that you are responding to the OP, which I am not sure is correct. Sorry if I am misunderstanding.

**Comment by realazthat** [comment]
> The thing that I was confused about was what exactly was meant by your question becuase the web's file system API cannot be used on the command line:
> 
> >

Right, which is why i don't understand how your solution is a solution.

**Comment by realazthat** [comment]
To be clear: I am trying to use excalidraw in my book, which uses LaTeX tooling. For graphviz, I use dot, and it takes my `.dot` files, and turns them into SVGs. For other diagram formats I use other tools. For excalidraw, I have a `.excalidraw` file stored locally, and in git, but i have no way to generate the SVGs without doing that by hand which is quite labor intensive. "Use the website" is not a tool. Yes, I can use the website on any platform, but the OP is addressing usage as a tool, for _automated_ conversion, so it can be used in an automated workflow.

**Comment by eunjae-lee** [comment]
> To be clear: I am trying to use excalidraw in my book, which uses LaTeX tooling. For graphviz, I use dot, and it takes my `.dot` files, and turns them into SVGs. For other diagram formats I use other tools. For excalidraw, I have a `.excalidraw` file stored locally, and in git, but i have no way to generate the SVGs without doing that by hand which is quite labor intensive. "Use the website" is not a tool. Yes, I can use the website on any platform, but the OP is addressing usage as a tool, for _automated_ conversion, so it can be used in an automated workflow.

I asked for two features which may have confused all the discussions :)

**Comment by realazthat** [comment]
Ah my bad, I think I get the confusion now.

**Comment by pseudo-su** [comment]
Might be an option now https://github.com/JonRC/excalirender



> The rendering pipeline reads `.excalidraw` JSON files and draws elements to a server-side canvas using the same libraries Excalidraw uses:
>
> - [Rough.js](https://roughjs.com/) for hand-drawn shapes
> - [perfect-freehand](https://github.com/steveruizok/perfect-freehand) for pressure-sensitive strokes
> - [node-canvas](https://github.com/Automattic/node-canvas) for server-side rendering

**Comment by wh1le** [comment]
Built a CLI that does this: https://github.com/wh1le/excalidraw-edit
```bash
npm i -g excalidraw-edit
excalidraw-edit drawing.excalidraw
```
Opens locally, auto-saves to disk, works offline.
